File: ios/loveaceios/ViewModel/ElectricityViewModel.swift
import Foundation

@MainActor @Observable
final class ElectricityViewModel {
    var isLoading = false
    var isLoadingOptions = false
    var buildings: [[String: String]] = []
    var floors: [[String: String]] = []
    var rooms: [[String: String]] = []
    var selectedBuilding: (code: String, name: String)?
    var selectedFloor: (code: String, name: String)?
    var selectedRoom: (code: String, name: String)?
    var electricityInfo: ElectricityInfo?
    var error: String?
    var boundRoomCode: String?
    var boundRoomDisplay: String?
    private var service: ISIMService?

    func initialize(service: ISIMService) {
        self.service = service
        boundRoomCode = UserDefaults.standard.string(forKey: "elec_room_code")
        boundRoomDisplay = UserDefaults.standard.string(forKey: "elec_room_display")
    }

    func autoLoad() {
        guard let code = boundRoomCode else { return }
        loadElectricityInfo(roomCode: code, displayText: boundRoomDisplay)
    }

    func loadBuildings() {
        guard let svc = service else { return }
        Task {
            isLoadingOptions = true; error = nil
            let result = await svc.getBuildings()
            buildings = result.data ?? []; isLoadingOptions = false
            if !result.success { error = result.error }
        }
    }

    func selectBuilding(code: String, name: String) {
        selectedBuilding = (code, name); selectedFloor = nil; selectedRoom = nil; floors = []; rooms = []
        guard let svc = service else { return }
        Task { isLoadingOptions = true; floors = (await svc.getFloors(buildingCode: code)).data ?? []; isLoadingOptions = false }
    }

    func selectFloor(code: String, name: String) {
        selectedFloor = (code, name); selectedRoom = nil; rooms = []
        guard let svc = service else { return }
        Task { isLoadingOptions = true; rooms = (await svc.getRooms(floorCode: code)).data ?? []; isLoadingOptions = false }
    }

    func selectRoom(code: String, name: String) { selectedRoom = (code, name) }

    func getService() -> ISIMService? { service }

    func bindRoom(code: String, display: String) {
        UserDefaults.standard.set(code, forKey: "elec_room_code")
        UserDefaults.standard.set(display, forKey: "elec_room_display")
        boundRoomCode = code; boundRoomDisplay = display
        loadElectricityInfo(roomCode: code, displayText: display)
    }

    func confirmBinding() {
        guard let building = selectedBuilding, let floor = selectedFloor, let room = selectedRoom else { return }
        let display = "\(building.name) \(floor.name) \(room.name)"
        bindRoom(code: room.code, display: display)
    }

    func loadElectricityInfo(roomCode: String, displayText: String? = nil) {
        guard let svc = service else { return }
        Task {
            isLoading = true; error = nil
            let result = await svc.getElectricityInfo(roomCode: roomCode, displayText: displayText)
            electricityInfo = result.data; isLoading = false
            if !result.success { error = result.error }
        }
    }

    func clearBinding() {
        UserDefaults.standard.removeObject(forKey: "elec_room_code")
        UserDefaults.standard.removeObject(forKey: "elec_room_display")
        boundRoomCode = nil; boundRoomDisplay = nil; electricityInfo = nil
        buildings = []; floors = []; rooms = []
        selectedBuilding = nil; selectedFloor = nil; selectedRoom = nil
    }
}

File: android/app/src/main/java/tech/loveace/appv3/ui/viewmodel/ElectricityViewModel.kt
package tech.loveace.appv3.ui.viewmodel

import android.app.Application
import android.content.Context
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import tech.loveace.appv3.data.model.ElectricityInfo
import tech.loveace.appv3.data.service.ISIMService

data class RoomSelection(val code: String, val name: String)

data class ElectricityUiState(
    val isLoading: Boolean = false,
    val buildings: List<Map<String, String>> = emptyList(),
    val floors: List<Map<String, String>> = emptyList(),
    val rooms: List<Map<String, String>> = emptyList(),
    val selectedBuilding: RoomSelection? = null,
    val selectedFloor: RoomSelection? = null,
    val selectedRoom: RoomSelection? = null,
    val electricityInfo: ElectricityInfo? = null,
    val error: String? = null,
    // 已绑定的房间
    val boundRoomCode: String? = null,
    val boundRoomDisplay: String? = null,
)

class ElectricityViewModel(application: Application) : AndroidViewModel(application) {
    private var service: ISIMService? = null
    private val _uiState = MutableStateFlow(ElectricityUiState())
    val uiState: StateFlow<ElectricityUiState> = _uiState.asStateFlow()

    private val prefs = application.getSharedPreferences("electricity_room", Context.MODE_PRIVATE)

    fun init(service: ISIMService) {
        this.service = service
        // 恢复已绑定的房间
        val code = prefs.getString("room_code", null)
        val display = prefs.getString("room_display", null)
        if (code != null && display != null) {
            _uiState.value = _uiState.value.copy(boundRoomCode = code, boundRoomDisplay = display)
        }
    }

    /** 自动加载：如果已绑定房间，直接查询电费 */
    fun autoLoad() {
        val code = _uiState.value.boundRoomCode ?: return
        val display = _uiState.value.boundRoomDisplay
        loadElectricityInfo(code, display)
    }

    fun loadBuildings() {
        val svc = service ?: return
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true, error = null)
            val result = svc.getBuildings()
            _uiState.value = if (result.success)
                _uiState.value.copy(isLoading = false, buildings = result.data ?: emptyList())
            else _uiState.value.copy(isLoading = false, error = result.error)
        }
    }

    fun selectBuilding(code: String, name: String) {
        _uiState.value = _uiState.value.copy(
            selectedBuilding = RoomSelection(code, name),
            selectedFloor = null, selectedRoom = null,
            floors = emptyList(), rooms = emptyList(),
        )
        val svc = service ?: return
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true)
            val result = svc.getFloors(code)
            _uiState.value = _uiState.value.copy(
                isLoading = false,
                floors = if (result.success) result.data ?: emptyList() else emptyList(),
            )
        }
    }

    fun selectFloor(code: String, name: String) {
        _uiState.value = _uiState.value.copy(
            selectedFloor = RoomSelection(code, name),
            selectedRoom = null, rooms = emptyList(),
        )
        val svc = service ?: return
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true)
            val result = svc.getRooms(code)
            _uiState.value = _uiState.value.copy(
                isLoading = false,
                rooms = if (result.success) result.data ?: emptyList() else emptyList(),
            )
        }
    }

    fun selectRoom(code: String, name: String) {
        _uiState.value = _uiState.value.copy(selectedRoom = RoomSelection(code, name))
    }

    /** 确认绑定房间并查询电费 */
    fun confirmBinding() {
        val building = _uiState.value.selectedBuilding ?: return
        val floor = _uiState.value.selectedFloor ?: return
        val room = _uiState.value.selectedRoom ?: return
        val display = "${building.name} ${floor.name} ${room.name}"
        // 保存绑定
        prefs.edit().putString("room_code", room.code).putString("room_display", display).apply()
        _uiState.value = _uiState.value.copy(boundRoomCode = room.code, boundRoomDisplay = display)
        loadElectricityInfo(room.code, display)
    }

    fun loadElectricityInfo(roomCode: String, displayText: String? = null) {
        val svc = service ?: return
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true, error = null)
            val result = svc.getElectricityInfo(roomCode, displayText)
            _uiState.value = if (result.success)
                _uiState.value.copy(isLoading = false, electricityInfo = result.data)
            else _uiState.value.copy(isLoading = false, error = result.error)
        }
    }

    /** 清除绑定 */
    fun clearBinding() {
        prefs.edit().clear().apply()
        _uiState.value = ElectricityUiState()
    }
}
